src/redisgears_python.c

```diff
@@ -2093,6 +2093,90 @@ static PyObject* createTensorFromValues(PyObject *cls, PyObject *args){
+        return NULL;
+    }
+    PyObject* keyName = PyTuple_GetItem(args, 0);
+    if(!PyUnicode_Check(keyName)){
+        PyErr_SetString(GearsError, "key name argument must be a string");
+        return NULL;
+    }
+    PyTensor* pyt = (PyTensor*)PyTuple_GetItem(args, 1);
+    if(!PyObject_IsInstance((PyObject*)pyt, (PyObject*)&PyTensorType)){
+        PyErr_SetString(GearsError, "Given argument is not of type PyTensor");
+        return NULL;
+    }
+
+    RedisModuleType *ai_tensorType = RedisAI_TensorRedisType();
+    size_t len;
+    const char* keyNameCStr = PyUnicode_AsUTF8AndSize(keyName, &len);
+
+    RedisModuleCtx* rctx = RedisModule_GetThreadSafeContext(NULL);
+    RedisModuleString *keyNameStr = RedisModule_CreateString(rctx, keyNameCStr, len);
+    LockHandler_Acquire(rctx);
+    RedisModuleKey *key = RedisModule_OpenKey(rctx,keyNameStr, REDISMODULE_WRITE);
+    // Set a shllow copy of the tensor.
+    RedisModule_ModuleTypeSetValue(key, ai_tensorType, RedisAI_TensorGetShallowCopy(pyt->t));
+    LockHandler_Release(rctx);
+    RedisModule_FreeString(rctx, keyNameStr);
+    RedisModule_FreeThreadSafeContext(rctx);
+    return PyLong_FromLong(1);
+}
+
+static PyObject* getTensorFromKey(PyObject *cls, PyObject *args) {
+     verifyRedisAILoaded();
+    // Input validation: 1 argument. args[0]: string.
+    if(PyTuple_Size(args) != 1){
+        PyErr_SetString(GearsError, "Wrong number of arguments given to setTensorInKey");
+        return NULL;
+    }
+    PyObject* keyName = PyTuple_GetItem(args, 0);
+    if(!PyUnicode_Check(keyName)){
+        PyErr_SetString(GearsError, "key name argument must be a string");
+        return NULL;
+    }
+
+    RedisModuleType *ai_tensorType = RedisAI_TensorRedisType();
+    size_t len;
+    const char* keyNameCStr = PyUnicode_AsUTF8AndSize(keyName, &len);
+
+    RedisModuleCtx* rctx = RedisModule_GetThreadSafeContext(NULL);
+    RedisModuleString *keyNameStr = RedisModule_CreateString(rctx, keyNameCStr, len);
+    LockHandler_Acquire(rctx);
+    RedisModuleKey *key = RedisModule_OpenKey(rctx,keyNameStr, REDISMODULE_READ);
+    // Empty key.
+    if (RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+        RedisModule_CloseKey(key);
+        LockHandler_Release(rctx);
+        RedisModule_FreeString(rctx, keyNameStr);
+        RedisModule_FreeThreadSafeContext(rctx);
+        PyErr_SetString(GearsError, "ERR tensor key is empty");
+        return NULL;
+    }
+    // No tensor key.
+    if (RedisModule_ModuleTypeGetType(key) != ai_tensorType) {
+        RedisModule_CloseKey(key);
+        LockHandler_Release(rctx);
+        RedisModule_FreeString(rctx, keyNameStr);
+        RedisModule_FreeThreadSafeContext(rctx);
+        PyErr_SetString(GearsError, "ERR key is not a tensor");
+        return NULL;
+    }
+    // Read tensor and free locks.
+    RAI_Tensor *tensor = RedisModule_ModuleTypeGetValue(key);
+    LockHandler_Release(rctx);
```

Comment:
You should only free the lock after you get the ShallowCopy, otherwise someone might delete the key (and free the tensor) before you get the ShallowCopy.